src/libraries/Native/Unix/System.Native/pal_io.c

```diff
@@ -1288,3 +1297,31 @@ int32_t SystemNative_LChflagsCanSetHiddenFlag(void)
     return false;
 #endif
 }
+
+#ifdef __sun
+
+int32_t SystemNative_ReadProcessStatusInfo(pid_t pid, ProcessStatus* processStatus)
+{
+    char statusFilename[64];
+    snprintf(statusFilename, sizeof(statusFilename), "/proc/%d/psinfo", pid);
+
+    intptr_t fd;
+    while ((fd = open(statusFilename, O_RDONLY)) < 0 && errno == EINTR);
+    if (fd < 0)
+    {
+        return 0;
+    }
+
+    psinfo_t status;
+    int result = Common_Read(fd, &status, sizeof(psinfo_t));
+    while (close(fd) < 0 && errno == EINTR);
```

Comment:
It's generally discouraged to retry close.  Is that not the case on SunOS?

Reply:
I found a similar implementation in clr's pal (and used its compact form): https://github.com/dotnet/runtime/blob/6072e4d3a7a2a1493f514cdf4be75a3d56580e84/src/coreclr/src/pal/src/sharedmemory/sharedmemory.cpp#L294-L298

Reply:
I have simplified it to just call `close(fd)` for now; in case the call is failed with `EINTR`, it will leak the handle. The current documentation (https://illumos.org/man/2/close) says that in that case the state of fd is unspecified.

```
       If close() is interrupted by a signal that is to be caught, it will
       return -1 with errno set to EINTR and the state of fildes is unspecified.
       If an I/O error occurred while reading from or writing to the file system
       during close(), it returns -1, sets errno to EIO, and the state of fildes
       is unspecified.
```

but it is not clear whether retrying would make things better or cause some other unexpected error (sigsegv).